### PR TITLE
Refactor is_failover_possible()

### PR DESCRIPTION
**Pull request by hughcapet**
Move all the members filtering inside `is_failover_possible()`. Rename "members list" to "candidates list"

### CODE REVIEW

File: patroni/ha.py [hunk 1]
```diff
@@ -1938,23 +1933,28 @@ def get_remote_member(self, member: Union[Leader, Member, None] = None) -> Remot
...
-        :param check_sync: if ``True``, also check against the sync key members
+        :param check_synchronous: if ``True``, also check against the sync key members.
+        :param exclude_candidate: if ``True``, exclude :attr:`failover.candidate` from the candidates.
 
-        :returns: a list of ``Member`` ojects or an empty list if there is no candidate available
+        :returns: a list of :class:`Member` ojects or an empty list if there is no candidate available.
         """
-        failover = self.cluster.failover
-        if check_sync:
+        def is_eligible(node_name: str, check_synchronous: bool, exclude_candidate: bool) -> bool:
```
Comment (CyberDem0n) [inline patroni/ha.py:1950]: We better pass `Member` object, so we can add additional checks for api_url and that nofailover isn't set.
check_synchronous and check_synchronous are not needed here.
```suggestion
        def is_eligible(node: Member) -> bool:
```

File: patroni/ha.py [hunk 2]
```diff
@@ -1938,23 +1933,28 @@ def get_remote_member(self, member: Union[Leader, Member, None] = None) -> Remot
...
-            # every sync_standby or the candidate specified if is in sync_standbys
-            return [m for m in self.cluster.members
-                    if self.cluster.sync.matches(m.name)
-                    and (not failover or not failover.candidate or m.name == failover.candidate)]
-        else:
-            # every member or the candidate specified
-            return [m for m in self.cluster.members
-                    if not failover or not failover.candidate or m.name == failover.candidate]
+            failover = self.cluster.failover
+            exclude = [self.state_handler.name] + ([failover.candidate] if failover and exclude_candidate else [])
```
Comment (CyberDem0n) [inline patroni/ha.py:1953]: These two lines could be moved outside the function (to the outer method).

File: patroni/ha.py [hunk 3]
```diff
@@ -1938,23 +1933,28 @@ def get_remote_member(self, member: Union[Leader, Member, None] = None) -> Remot
...
+            if check_synchronous and self.is_synchronous_mode() and not self.cluster.sync.is_empty and \
+               not self.cluster.sync.matches(node_name):
```
Comment (CyberDem0n) [inline patroni/ha.py:1955]:
```suggestion
               not self.cluster.sync.matches(node.name):
```

File: patroni/ha.py [hunk 4]
```diff
@@ -1938,23 +1933,28 @@ def get_remote_member(self, member: Union[Leader, Member, None] = None) -> Remot
...
+                return False
+            return node_name not in exclude and \
```
Comment (CyberDem0n) [inline patroni/ha.py:1957]: The following will help to get rid of https://github.com/zalando/patroni/pull/2804/files#diff-b6a3a0ef64783a6d32feb908d74a3a009e9f25008b8ca5313c9c9747e3b74a63R925
```suggestion
            return node.name not in exclude and and not node.nofailover and node.api_url \
```

File: patroni/ha.py [hunk 5]
```diff
@@ -1938,23 +1933,28 @@ def get_remote_member(self, member: Union[Leader, Member, None] = None) -> Remot
...
+                (not failover or not failover.candidate or node_name == failover.candidate)
+
+        return list(filter(lambda m: is_eligible(m.name, check_synchronous, exclude_candidate), self.cluster.members))
```
Comment (CyberDem0n) [inline patroni/ha.py:1960]:
```suggestion
        return list(filter(is_eligible, self.cluster.members))
```

File: patroni/ha.py [hunk 6]
```diff
@@ -1877,7 +1872,7 @@ def _on_shutdown(checkpoint_location: int) -> None:
                     # location, we can remove the leader key and allow them to start leader race.
 
                     # for a manual failover/switchover with a candidate, we should check the requested candidate only
-                    if self.is_failover_possible(self.get_failover_candidates(), cluster_lsn=checkpoint_location):
+                    if self.is_failover_possible(False, checkpoint_location):
```
Comment (CyberDem0n) [inline patroni/ha.py:1875]: It seems like you changed the logic. In this place we have to do same checks as like normal failover happens:
1. if synchronous replication is enable - filter out non-sync nodes
2. if there is a failover.candidate - check that only the candidate received the checkpoint_location
3. and so on
 
```suggestion
                    if self.is_failover_possible(cluster_lsn=checkpoint_location):
```

File: patroni/ha.py [hunk 7]
```diff
@@ -1214,14 +1216,7 @@ def process_manual_failover_from_leader(self) -> Optional[str]:
...
                 else:
-                    if self.is_synchronous_mode():
-                        members = self.get_failover_candidates(check_sync=True)
-                        if failover.candidate and not members:
-                            logger.warning('Failover candidate=%s does not match with sync_standbys=%s',
-                                           failover.candidate, self.cluster.sync.sync_standby)
-                    else:
-                        members = self.get_failover_candidates()
-                    if self.is_failover_possible(members, False):  # check that there are healthy members
+                    if self.is_failover_possible(self.is_synchronous_mode(), exclude_candidate=False):
```
Comment (CyberDem0n) [inline patroni/ha.py:1219]: Something doesn't feel right here.

**Review by CyberDem0n — COMMENTED** [review]
@hughcapet please don't apply blindly any suggestions and double check everything

File: patroni/ha.py [hunk 8]
```diff
@@ -904,26 +904,31 @@ def _is_healthiest_node(self, members: Collection[Member], check_replication_lag
                         logger.info('Ignoring the former leader being ahead of us')
         return True
 
-    def is_failover_possible(self, members: List[Member], check_synchronous: Optional[bool] = True,
-                             cluster_lsn: Optional[int] = 0) -> bool:
-        """Checks whether one of the members from the list can possibly win the leader race.
+    def is_failover_possible(self, check_synchronous: bool = True, cluster_lsn: int = 0,
+                             exclude_candidate: bool = False) -> bool:
```
Comment (CyberDem0n) [inline patroni/ha.py:908]:
```suggestion
    def is_failover_possible(self, *, cluster_lsn: int = 0, exclude_candidate: bool = False) -> bool:
```

File: patroni/ha.py [hunk 9]
```diff
@@ -904,26 +904,31 @@ def _is_healthiest_node(self, members: Collection[Member], check_replication_lag
...
+        """Checks whether any of the cluster members is allowed to promote and is healthy enough for that.
 
-        :param members: list of members to check
         :param check_synchronous: consider only members that are known to be listed in /sync key when sync replication.
```
Comment (CyberDem0n) [inline patroni/ha.py:911]:
```suggestion
```

File: patroni/ha.py [hunk 10]
```diff
@@ -904,26 +904,31 @@ def _is_healthiest_node(self, members: Collection[Member], check_replication_lag
...
-        :param cluster_lsn: to calculate replication lag and exclude member if it is laggin
-        :returns: `True` if there are members eligible to be the new leader
+        :param cluster_lsn: to calculate replication lag and exclude member if it is lagging.
+        :param exclude_candidate: if ``True``, exclude :attr:`failover.candidate` from the members list against which
+            the failover possibility checks are run.
+        :returns: `True` if there are members eligible to become the new leader.
         """
+        candidates = self.get_failover_candidates(check_synchronous, exclude_candidate)
```
Comment (CyberDem0n) [inline patroni/ha.py:917]:
```suggestion
        candidates = self.get_failover_candidates(exclude_candidate)
```

File: patroni/ha.py [hunk 11]
```diff
@@ -904,26 +904,31 @@ def _is_healthiest_node(self, members: Collection[Member], check_replication_lag
...
+
+        if check_synchronous and self.cluster.failover and self.cluster.failover.candidate and not candidates:
```
Comment (CyberDem0n) [inline patroni/ha.py:919]:
```suggestion
        if self.is_synchronous_mode() and not self.cluster.sync.is_empty \
                and self.cluster.failover and self.cluster.failover.candidate and not candidates:
```

File: patroni/ha.py [hunk 12]
```diff
@@ -904,26 +904,31 @@ def _is_healthiest_node(self, members: Collection[Member], check_replication_lag
...
+                           self.cluster.failover.candidate, self.cluster.sync.sync_standby)
+
         ret = False
         cluster_timeline = self.cluster.timeline
-        members = [m for m in members if m.name != self.state_handler.name and not m.nofailover and m.api_url]
-        if check_synchronous and self.is_synchronous_mode() and not self.cluster.sync.is_empty:
-            members = [m for m in members if self.cluster.sync.matches(m.name)]
-        if members:
-            for st in self.fetch_nodes_statuses(members):
+        candidates = [m for m in candidates if not m.nofailover and m.api_url]
```
Comment (CyberDem0n) [inline patroni/ha.py:925]:
```suggestion
```

File: patroni/ha.py [hunk 13]
```diff
@@ -1109,9 +1113,7 @@ def on_shutdown(checkpoint_location: int) -> None:
             # It could happen if Postgres is still archiving the backlog of WAL files.
             # If we know that there are replicas that received the shutdown checkpoint
             # location, we can remove the leader key and allow them to start leader race.
-
-            # for a manual failover/switchover with a candidate, we should check the requested candidate only
-            if self.is_failover_possible(self.get_failover_candidates(), cluster_lsn=checkpoint_location):
+            if self.is_failover_possible(False, checkpoint_location):
```
Comment (CyberDem0n) [inline patroni/ha.py:1116]:
```suggestion
            if self.is_failover_possible(cluster_lsn, checkpoint_location):
```

File: patroni/ha.py (hunk 7 above)
Comment (CyberDem0n) [inline patroni/ha.py:1219]:
```suggestion
                    if self.is_failover_possible():
```

File: patroni/ha.py [hunk 14]
```diff
@@ -1938,23 +1933,28 @@ def get_remote_member(self, member: Union[Leader, Member, None] = None) -> Remot
         name = member.name if member else 'remote_member:{}'.format(uuid.uuid1())
         return RemoteMember(name, data)
 
-    def get_failover_candidates(self, check_sync: bool = False) -> List[Member]:
-        """Return list of candidates for either manual or automatic failover.
+    def get_failover_candidates(self, check_synchronous: bool, exclude_candidate: bool) -> List[Member]:
```
Comment (CyberDem0n) [inline patroni/ha.py:1936]:
```suggestion
    def get_failover_candidates(self, exclude_candidate: bool) -> List[Member]:
```

File: patroni/ha.py [hunk 15]
```diff
@@ -1938,23 +1933,28 @@ def get_remote_member(self, member: Union[Leader, Member, None] = None) -> Remot
...
 
-        Mainly used to later be passed to ``Ha.is_failover_possible()``.
+        Exclude non-sync members when in synchronous mode, the current node (its checks are always performed earlier)
+        and the candidate if required. If failover candidate exclusion is not requested and a candidate is specified
+        in the /failover key, return the candidate only.
+        The result is further evaluated in the caller :func:`Ha.is_failover_possible` to check if any member is actually
+        healthy enough and is allowed to poromote.
 
-        :param check_sync: if ``True``, also check against the sync key members
+        :param check_synchronous: if ``True``, also check against the sync key members.
```
Comment (CyberDem0n) [inline patroni/ha.py:1945]:
```suggestion
```

File: patroni/ha.py (hunk 2 above)
Comment (CyberDem0n) [inline patroni/ha.py:1953]:
```suggestion
```

**Review by CyberDem0n — COMMENTED** [review]
And unit tests:
```diff
diff --git a/tests/test_ha.py b/tests/test_ha.py
index d3a75ed3..6eb38204 100644
--- a/tests/test_ha.py
+++ b/tests/test_ha.py
@@ -748,7 +748,6 @@ class TestHa(PostgresInit):
         self.p.is_leader = true
         self.ha.has_lock = true
         self.ha.is_synchronous_mode = true
-        self.ha.is_failover_possible = false
         self.ha.process_sync_replication = Mock()
         self.ha.cluster = get_cluster_initialized_with_leader(Failover(0, self.p.name, 'a', None), (self.p.name, None))
         self.assertEqual('no action. I am (postgresql0), the leader with the lock', self.ha.run_cycle())
```

File: patroni/ha.py (hunk 4 above)
Comment (hughcapet) [reply patroni/ha.py:1957]: with this I feel like we start to mix up "eligibility" and "healthiness"
maybe we need to define the terms then

**Comment by CyberDem0n** [comment]
:+1:

**Comment by hughcapet** [comment]
👍

**Pull request merged**